docs/design/coreclr/botr/memory-model.md

```diff
@@ -0,0 +1,248 @@
+
+# CLR memory model
+
+## ECMA vs. CLR memory models.
+ECMA 335 standard defines a very weak memory model. After two decades the desire to have a flexible model did not result in considerable benefits due to hardware being more strict. On the other hand programming against ECMA model requires extra complexity to handle scenarios that are hard to comprehend and not possible to test.
+
+In the course of multiple releases CLR implementation settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the CLR runtime in its current implementation with expectation of that being carried to future releases.
+
+## Alignment
+When managed by CLR runtime, variables of built-in primitive types are *properly aligned* according to the data type size. This applies to both heap and stack allocated memory.
```

Comment:
What about members of structs?

Reply:
Fields are variables. 

Regardless how you get something that has, for example, type `int`, it will be aligned accordingly. The runtime (i.e. JIT, GC) will ensure alignment for all variables under runtime's control.

Reply:
`FieldOffsetAttribute` could possibly introduce unaligned fields.

Reply:
Arguably, by using `FieldOffsetAttribute` you are not letting the runtime to manage locations of the fields. 
I will add a note about `FieldOffsetAttribute` as something that may cause unaligned variables.